sqlglot/optimizer/simplify.py

```diff
@@ -374,47 +373,88 @@ def absorb_and_eliminate(expression, root=True):
     return expression
 
 
-def propagate_constants(expression, root=True):
+# Type alias for a map of columns to their constant value
+ConstantMapping = t.Dict[exp.Column, t.Tuple[int, exp.Expression]]
+
+
+def _get_constant_mapping(ops: t.Iterable[exp.Expression]) -> ConstantMapping:
     """
-    Propagate constants for conjunctions in DNF:
+    Extract a constant mapping from `ops`
+    """
+    constant_mapping: ConstantMapping = {}
+
+    for op in ops:
+        if isinstance(op, exp.EQ):
+            l, r = op.left, op.right
+
+            # TODO: create a helper that can be used to detect nested literal expressions such
+            # as CAST(123456 AS BIGINT), since we usually want to treat those as literals too
+            if isinstance(l, exp.Column) and isinstance(r, exp.Literal):
+                pass
+            elif isinstance(r, exp.Column) and isinstance(l, exp.Literal):
+                l, r = r, l
+            else:
+                continue
+
+            constant_mapping[l] = (id(l), r)
+
+    return constant_mapping
+
+
+def _replace_column(column: exp.Column, constant_mapping: ConstantMapping) -> None:
+    """
+    Potentially replace `column` with a constant from `constant_mapping`
+    """
+    parent = column.parent
+    column_id, constant = constant_mapping.get(column) or (None, None)
+    if (
+        column_id is not None
+        and constant is not None
+        and id(column) != column_id
+        and not (isinstance(parent, exp.Is) and isinstance(parent.expression, exp.Null))
+    ):
+        column.replace(constant.copy())
+
+
+def propagate_constants(expression: exp.Expression, root: bool = True) -> exp.Expression:
+    """
+    Propagate constants for conjunctions:
 
     SELECT * FROM t WHERE a = b AND b = 5 becomes
     SELECT * FROM t WHERE a = 5 AND b = 5
 
     Reference: https://www.sqlite.org/optoverview.html
+
+    This also propagates constants into projections:
+
+    SELECT a FROM x WHERE a = 1 becomes
+    SELECT 1 FROM x WHERE a = 1
     """
+    if isinstance(expression, exp.Select):
+        where = expression.args.get("where")
+        if not where:
+            return expression
 
-    if (
-        isinstance(expression, exp.And)
-        and (root or not expression.same_parent)
-        and sqlglot.optimizer.normalize.normalized(expression, dnf=True)
```

Comment:
Why did we remove the DNF check for `expression`?

Reply:
I don't think you actually wanted DNF? Shouldn't it be CNF?
Here, we check the top level ANDs for equalities and don't traverse into any ORs when replacing. I think its effectively the same thing.

Reply:
> I don't think you actually wanted DNF? Shouldn't it be CNF?

I think DNF was right here - you want to aggregate all conjunction-only terms and run the constant propagation rule inside them. So for example if you have `(A Λ B Λ C) V (D Λ E Λ F) V ...`, which is in DNF, you can propagate constants in each of the sub-terms.

Reply:
>  Here, we check the top level ANDs for equalities and don't traverse into any ORs when replacing. I think its effectively the same thing.

I see. I'd need to think about this a bit more to understand the implications, e.g. in inputs where the con/disjunctions are not explicitly grouped (i.e. A AND B OR C, stuff like that). At the time I think the DNF check felt more safe, which is why I did it that way.

Reply:
Ah, that's a good point. I forget the AST doesn't actually have correct precedence

Reply:
Er, or does it? I'll have to think through this...

Reply:
The AST _does_ encode the correct precedence in general, but different dialects may prioritize the parsing of `AND` and `OR` differently, i.e. assign different precedence levels to them when not grouped (though I don't have an example for this right now, but can check)